clients/da-vinci-client/src/main/java/com/linkedin/davinci/consumer/VeniceAfterImageConsumerImpl.java

```diff
@@ -1,17 +1,210 @@
+          }
+
+          // poll until we get EOP for all partitions
+          while (true) {
+            polledResults = consumerAdapter.poll(5000L);
+            // Loop through all polled messages
+            for (Map.Entry<PubSubTopicPartition, List<PubSubMessage<KafkaKey, KafkaMessageEnvelope, Long>>> entry: polledResults
+                .entrySet()) {
+              PubSubTopicPartition pubSubTopicPartition = entry.getKey();
+              List<PubSubMessage<KafkaKey, KafkaMessageEnvelope, Long>> messageList = entry.getValue();
+              for (PubSubMessage<KafkaKey, KafkaMessageEnvelope, Long> message: messageList) {
+                if (message.getKey().isControlMessage()) {
+                  ControlMessage controlMessage = (ControlMessage) message.getValue().getPayloadUnion();
+                  ControlMessageType controlMessageType = ControlMessageType.valueOf(controlMessage);
+                  if (controlMessageType.equals(ControlMessageType.END_OF_PUSH)) {
+                    // note down the partition and offset and mark that we've got the thing
+                    endOfPushConsumedPerPartitionMap.put(pubSubTopicPartition.getPartitionNumber(), true);
+                    VeniceChangeCoordinate coordinate = new VeniceChangeCoordinate(
+                        pubSubTopicPartition.getPubSubTopic().getName(),
+                        new ApacheKafkaOffsetPosition(message.getOffset()),
+                        pubSubTopicPartition.getPartitionNumber());
+                    checkpoints.add(coordinate);
+                    Set<Integer> unsubSet = new HashSet<>();
+                    unsubSet.add(pubSubTopicPartition.getPartitionNumber());
+                    internalSeekConsumer.get().unsubscribe(unsubSet);
+                    // No need to look at the rest of the messages for this partition that we might have polled
+                    break;
+                  }
+                }
+              }
+            }
+            if (endOfPushConsumedPerPartitionMap.values().stream().allMatch(e -> e)) {
+              // We polled all EOP messages, stop polling!
+              break;
+            }
+          }
+          this.seekToCheckpoint(checkpoints).get();
+        } catch (InterruptedException | ExecutionException e) {
+          throw new VeniceException(
+              "Seek to End of Push Failed for store: " + storeName + " partitions: " + partitions.toString(),
+              e);
+        }
+      }
+      return null;
+    });
+  }
+
+  @Override
+  protected CompletableFuture<Void> internalSeek(
+      Set<Integer> partitions,
+      PubSubTopic targetTopic,
+      SeekFunction seekAction) {
+    if (!versionSwapDetectionThread.isAlive()) {
+      versionSwapDetectionThread.start();
+    }
+    return super.internalSeek(partitions, targetTopic, seekAction);
+  }
+
+  private class VersionSwapDetectionThread extends Thread {
+    VersionSwapDetectionThread() {
+      super("Version-Swap-Detection-Thread");
```

Comment:
So we might have 1 thread per AfterImageConsumer, is it possible we will have multiple AfterImageConsumer for one store? If so, we would better to add more informative name to this thread. This better carry the informatition carried from Factory creating consumer provided by customer.

Reply:
I'm not sure I can think of a scheme that would be especially helpful other then adding a random number at the end of the name.

Reply:
Added thread id to the loggers.  I think that'd be a better approach in all cases because then the id would match up in thread dumps.